[Tay Yi Hsuen] iP

yhtMinceraft1010X opened the pull request.

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

yhtMinceraft1010X committed:
-Added minimalist functionality to my version of the Duke chatbot, codenamed "Meme Man".
-Updated the README.md user guide to introduce Meme Man and describe "bye" command functionality.

yhtMinceraft1010X committed:
Level 2 changes:
-Removed echo functionality of Meme Man.
-Added functionality for Meme Man to note down tasks and display them when prompted.
-Update README.md user guide to reflect the above changes.

yhtMinceraft1010X committed:
Level 3 changes:
-Added functionality for Meme Man to keep track of whether tasks are done or not.
-Edited the logic for adding tasks. You are now required to put an add command before your task description.
-Update README.md user guide to reflect the above changes.

yhtMinceraft1010X committed:
Level 4 changes:
-Divided tasks into 3 types: ToDo, Deadline, Event.
-"add" command is no longer accepted. Use "todo", "deadline" and "event" commands to add ToDo, Deadline and Event tasks respectively.
-Removed spacing in front of task description for done and undone valid outputs.
-Added 2 easter eggs related to other Surreal Memes characters.
-Updated README.md user guide to reflect the above changes.

yhtMinceraft1010X committed:
Testing Automation:
-Added a test case to input.txt
-Slightly changed code inside Duke.java for formatting purposes

yhtMinceraft1010X committed:
Level 5 changes:
-Added Meme Man responses for exceptional user inputs using internal exceptions
-Updated README.md user guide to reflect above changes
-Reorganised and expanded number of test cases

yhtMinceraft1010X committed:
Level 6 changes:
-Added functionality for deleting tasks
-Redid exceptions for commands with number arguments (E.g. done 2, undone 3). They can now handle empty arguments and invalid arguments
-Updated the README.md user guide to reflect above changes
-Reorganised the user guide to better differentiate between task commands and other commands as well as command usage vs outputs
-Added more test cases to test delete functionality and edited test case 5 to account for above changes

yhtMinceraft1010X committed:
Reorganise code segments and rename program

Many methods for handling input and tasks are bundled together in the chatbot, which retains its legacy name Duke from the start.

However, having all the methods together in one place makes it difficult to find a method to debug. Also, keeping the original name is confusing.

Let's,
* separate the methods into different management classes
* rename the main program class to SurrealChat

Separating the methods into different management classes based on roles is preferable to keeping them together because it would be easier to find a method to debug. The SurrealChat name also eliminates confusion with regards to program name.

yhtMinceraft1010X committed:
Add save/load functionality for tasks

SurrealChat does not keep track of tasks between each run.

This causes time wastage in initialising tasks again.

Let's give a save/load functionality for tasks to be saved into a file and loaded from that file in the next run.

Using a file also allows one to keep track of tasks outside of the program.

yhtMinceraft1010X committed:
Reformat date as LocalDate

SurrealChat currently only takes in date as a string.

This makes sorting tasks according to date difficult.

Let's put the date as a LocalDate object instead.

This is preferable as it would be easier to compare dates.

yhtMinceraft1010X committed:
Merge branch 'branch-Level-7'

yhtMinceraft1010X committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/DeadlineTask.java
#	src/main/java/EventTask.java
#	src/main/java/TaskManagement.java
#	text-ui-test/Test_1/EXPECTED_OK_1.TXT
#	text-ui-test/Test_1/input1.txt
#	text-ui-test/Test_1/runtest1.bat
#	text-ui-test/Test_1/runtest1.sh
#	text-ui-test/Test_2/runtest2.bat
#	text-ui-test/Test_2/runtest2.sh
#	text-ui-test/Test_3/runtest3.bat
#	text-ui-test/Test_3/runtest3.sh
#	text-ui-test/Test_4/EXPECTED_OK_4.TXT
#	text-ui-test/Test_4/input4.txt
#	text-ui-test/Test_4/runtest4.bat
#	text-ui-test/Test_4/runtest4.sh
#	text-ui-test/Test_5/EXPECTED_OK_5.TXT
#	text-ui-test/Test_5/input5.txt
#	text-ui-test/Test_5/runtest5.bat
#	text-ui-test/Test_5/runtest5.sh
#	text-ui-test/Test_6/EXPECTED_OK_6.TXT
#	text-ui-test/Test_6/input6.txt
#	text-ui-test/Test_6/runtest6.bat
#	text-ui-test/Test_6/runtest6.sh
#	text-ui-test/Test_7/EXPECTED_OK_7.TXT
#	text-ui-test/Test_7/input7.txt
#	text-ui-test/Test_7/runtest7.bat
#	text-ui-test/Test_7/runtest7.sh
#	text-ui-test/Test_8/EXPECTED_OK_8.TXT
#	text-ui-test/Test_8/input8.txt
#	text-ui-test/Test_8/runtest8.bat
#	text-ui-test/Test_8/runtest8.sh

Resolution:
* DeadlineTask.java: Keep master, but modify deadline type in constructor.
* EventTask.java: Keep master, but modify event type in constructor.
* TaskManagement.java: Keep branch-Level-8, but add isDone variable behind.
* text-ui-test files: No actual conflicts found on manual inspection.

yhtMinceraft1010X committed:
Update .gitignore

Ignore a file under Git.

yhtMinceraft1010X committed:
Compartmentalise into classes

File operations as well as output and error print statements are scattered around the code.

However, these make it harder to modify print output and file operations as one would have to jump from class to class.

Let's,
* add a UserOutput class to handle print-related code
* add a FileManagement class to handle file operations
* add a TaskParser class to act as an intermediary for processing of tasks between IO side and TaskManagement side
* add a Pair class to store a command and an associated generic type item for further processing in UserOutput
* rename InputManagement to UserInput
* remove text-ui-test cases temporarily

Compartmentalising tasks makes it easier to search for the relevant code segment to modify and debug. Renaming InputManagement to UserInput also emphasises on the methods being used for user commands. In addition, removing the test cases for now allows for re-working of test cases without worrying about main code and vice versa.

yhtMinceraft1010X committed:
Package code

yhtMinceraft1010X committed:
Add JUnit tests

yhtMinceraft1010X committed:
Package into jar file

yhtMinceraft1010X committed:
Add JavaDoc comments

Most classes and methods did not come with JavaDoc comments.

This could lead to confusion on the purposes of these classes and methods.

Let's add JavaDoc comments to all non-private classes and methods.

This is preferable as classes and methods can be understood more easily.

yhtMinceraft1010X committed:
Update code to meet standards

Following standards improves readability of code by others.

Let's,
* unindent cases under switch statements
* add lines and comments between logical blocks in larger methods
* change public variables under classes to protected

yhtMinceraft1010X committed:
Add find task capability

Meme Man displays all tasks to the user.

This is impractical when the user only wishes to find certain tasks.

Let's add a find capability to SurrealChat that can take a keyword and find tasks with that keyword, regardless of upper or lower case.

This is preferable to reading through the entire list of tasks.

yhtMinceraft1010X committed:
Merge branch 'branch-A-JavaDoc'

yhtMinceraft1010X committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/DeadlineTask.java
#	src/main/java/EventTask.java
#	src/main/java/FileManagement.java

Resolution: Keep master

yhtMinceraft1010X committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/SurrealChat.java
#	src/main/java/TaskParser.java
#	src/main/java/UserOutput.java

Resolution: Keep changes from branch-Level-9. Also moved JavaDoc to match correct methods.

yhtMinceraft1010X committed:
Add exception handling for incorrect date format

The Deadline and Event tasks take in a date input in the form YYYY-MM-DD. If the input is incorrect, it would throw an uncaught exception.

This is not ideal as the program would terminate prematurely.

Let's,
* abstract out the code snippet used for parsing date in the methods for generating Deadline and Event tasks
* add exception handling for the incorrect date, along with a custom error message

This is preferable as code can continue to run even after the incorrect date format is keyed in and the message generated shares the same theme as other custom exception messages by SurrealChat. In addition, abstracting out the code makes changing the code easier if the date format is changed in the future.

yhtMinceraft1010X committed:
Update User Guide

The user guide presents the command and relevant inputs before the description of what the command does.

However, seeing the command and the arguments first can be an eyesore.

Let's,
* reorder the command and description. Reading from left to right, the description would come before the command.
* move the format of command and arguments from header to description.

This is preferable as the header is kept clean and readable.

yhtMinceraft1010X committed:
Add test cases

In more recent commits, there are only JUnit tests to test individual methods within the task subclasses.

Coverage of testing was poor. No system-wide testing was done.

Let's,
* reintroduce some of the text-ui-test cases.
* add a few more methods for JUnit tests.

This aims to improve testing coverage.

yhtMinceraft1010X committed:
Add edit task description functionality

Tasks are added with fixed descriptions.

Editing task descriptions is unfeasible since one has to delete the task and add a new task with the description.

Let's,
* add a functionality to edit task descriptions without deleting the task.
* make tasks immutable.
* add new JUnit tests to test editing of tasks and marking tasks as done or undone.

This is preferable as task descriptions can be conveniently edited. In addition, immutability helps protect the internal variables of the Task class from inadvertent editing.

yhtMinceraft1010X committed:
Rework code for commands and output

Code for handling user commands and file loading is mainly handled by TaskParser class. This also generates pairs of an internal tag and a generic second item of any type. The done and undone commands are also separate from each other.

This centralisation of code is confusing to look at. Having generics in UserOutput also results in compile warnings. A separate Done and Undone command not only duplicates code but also causes confusion should an undo/redo function be added in future.

Let's,
* reorganise code related to user commands into Command and EasterEgg classes and their respective subclasses.
* package the base .java files into actual packages based on package statements.
* generate the string outputs within the respective classes, rather than at the UserOutput class.
* move code for handling file operations from TaskParser class to TaskManagement class.
* consolidate done and undone command into done command to toggle a task status between done and undone.

This is preferable as editing code for an individual command can now be easily done at that command class. In addition, there is no need for the Pair class to carry unprocessed output for further processing at UserOutput and for the TaskParser class to handle user commands and file operations. Both the Pair class and TaskParser class can be removed. There is no more confusion between Undone task command marking a task as undone and a potential undo command for rolling back a certain change.

yhtMinceraft1010X committed:
Add help functionality

The README.md user guide is the only place to find guidance on how to use commands in SurrealChat.

This is inconvenient for people who need to use SurrealChat on the fly.

Let's add a help command that allows the user to read a brief description on how to use a command on the spot without having to pause to look up the user guide.

This is preferable as the user can quickly see the brief but important details on command usage and only need to refer to the user guide for examples and exception messages.

yhtMinceraft1010X committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

# Conflicts:
#	text-ui-test/runtest.sh

Resolution: Remove text-ui-test/runtest.sh as test files are restructured.

yhtMinceraft1010X committed:
Add Gradle support

Running of JUnit tests is done manually, test class by test class.

This can result in some failed test cases being missed as code changes if not all test classes are checked through. Javadoc errors may also be missed out.

Let's add Gradle support to the project.

This is preferable as all JUnit tests can be done together and Javadoc can be checked automatically.

yhtMinceraft1010X committed:
Merge branch 'branch-A-Gradle'

yhtMinceraft1010X committed:
Add Checkstyle

Code is checked manually to fit standards.

This is very inconvenient and is likely to get worse as more classes are added.

Let's,
* add checkstyle functionality.
* make use of checkstyle to find coding standard violations and correct them.

This is preferable as coding standard violations can be easily found.

yhtMinceraft1010X committed:
Merge branch 'branch-A-CheckStyle'

yhtMinceraft1010X committed:
Add GUI

The CLI is used for handling interactions between user and chatbot.

This has a dull aesthetic.

Let's,
* add a GUI with avatars.
* remove text-ui-tests.
* remove UserInput and UserOutput classes.

This is preferable as the user can now "see" who they are interacting with. In addition, text-ui-tests as well as UserInput and UserOutput classes are no longer necessary.

yhtMinceraft1010X committed:
Merge branch 'branch-Level-10'

RuiXiong2211 commented on src/main/java/SurrealChat.java:
```diff
@@ -0,0 +1,139 @@
+            List<String> fileLines = this.fileManagement.loadTaskFile();
+            this.taskManagement.parseFileLines(fileLines);
+        } catch (IOException e) {
+            return;
```
Perhaps you can write a comment to describe why the catch block returns nothing.

RuiXiong2211 commented on src/main/java/surrealchat/easteregg/EasterEgg.java:
```diff
@@ -0,0 +1,22 @@
+package surrealchat.easteregg;
+
+/**
+ * Base class for easter eggs.
```
Maybe you can give more description on what the easter egg class does

RuiXiong2211 commented on src/main/java/surrealchat/task/DeadlineTask.java:
```diff
@@ -0,0 +1,72 @@
+     * @return Deadline that is marked as done/undone.
+     */
+    public DeadlineTask markAsDone() {
+        return new DeadlineTask(this.getDescription(), this.deadline, !this.isDone);
```
An explanation on why you are creating new objects instead of altering the current object would be great!

RuiXiong2211 commented on src/main/java/surrealchat/command/DeleteCommand.java:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+
+    private boolean checkInvalidTaskNumber(int taskNumber, TaskManagement taskManagement) {
```
instead of checkInvalidTaskNumber, perhaps you can rename is to isInvalidTaskNumber.

RuiXiong2211 reviewed (commented):
Good Job! Code is neat and of high quality, method headers are well described and I could barely find any flaws in your code.

CharlesLee01 commented on src/main/java/SurrealChat.java:
```diff
@@ -0,0 +1,139 @@
+     * @return Output to be printed.
+     */
+    public String commandLogic(String inputString) {
+        String[] separatedWords = this.splitString(inputString);
```
I like the way you use appropriate naming for strings.

CharlesLee01 commented on src/main/java/SurrealChat.java:
```diff
@@ -0,0 +1,139 @@
+     * @return A string indicating that tasks are being saved.
+     */
+    public String saveFile() {
+        List<String> fileTaskList = this.taskManagement.convertTasksForFile();
```
I think you probably can name it as tasksFromFile.

CharlesLee01 commented on src/main/java/surrealchat/task/Task.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * Creates instance of Task object.
+     * @param description Description of task.
+     * @param type Type in terms of todo (denoted "T"), deadline (denoted "D"), event (denoted "E").
```
Good Job! Clear explanation of the symbol use for different tasks.

CharlesLee01 reviewed (commented):
Overall, you have done a good work trying to adhere to coding standard.

yhtMinceraft1010X committed:
Add time storage functionality and fix description spacing

Descriptions with spacing has spacing eliminated. The deadline and event tasks only keep track of date.

This is inconvenient to read and also imprecise for tracking purposes.

Let's,
* fix the spacing between words in task descriptions.
* add time storage to deadline and event tasks.
* add another fun easter egg.

This is preferable as the descriptions are readable and more precise task-keeping can be done.

yhtMinceraft1010X committed:
Add scronch all tasks functionality

The tasks are deleted one by one.

This can be inconvenient for users who want to delete everything in one go.

Let's add a scronch functionality to delete all tasks.

This is preferable to going through a long list of tasks and deleting them one by one.

yhtMinceraft1010X committed:
Add assertions

Let's replace some if guard conditions with assert statements.
This helps to reduce indentation levels in the code.

yhtMinceraft1010X committed:
Improve code quality

Let's,
* eliminate unnecessary "this." from classes.
* abstract out switch statements from longer methods.

This helps to improve readability of the code.

yhtMinceraft1010X committed:
Merge pull request #1 from yhtMinceraft1010X/branch-A-Assertions

Add assertions

yhtMinceraft1010X committed:
Merge branch 'master' of https://github.com/yhtMinceraft1010X/ip-1

* 'master' of https://github.com/yhtMinceraft1010X/ip-1:
  Add assertions

yhtMinceraft1010X committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/surrealchat/command/DeadlineCommand.java
#	src/main/java/surrealchat/command/DeleteCommand.java
#	src/main/java/surrealchat/command/DoneCommand.java
#	src/main/java/surrealchat/command/EditCommand.java
#	src/main/java/surrealchat/command/EventCommand.java
#	src/main/java/surrealchat/command/ToDoCommand.java

Resolution: Keep master, remove "this." across affected classes.

yhtMinceraft1010X committed:
Merge pull request #2 from yhtMinceraft1010X/branch-A-CodeQuality

Merge Branch-A-CodeQuality to master

yhtMinceraft1010X committed:
Merge branch 'master' of https://github.com/yhtMinceraft1010X/ip-1

* 'master' of https://github.com/yhtMinceraft1010X/ip-1:
  Improve code quality

yhtMinceraft1010X committed:
Reorganise exceptions and assertions

Let's,
* use a new SurrealException class for incorrect user inputs
* use assert statements to check assumptions regarding internal code and logic

yhtMinceraft1010X committed:
Update find functionality and add a stream

The find functionality returns a list of tasks in order of 1, 2, 3.

This confuses users when they need the actual task number for commands which require numbers.

Let's,
* edit the find functionality such that the actual task numbers are displayed in search results.
* move the internal workings of find from FindCommand to TaskManagement to enhance encapsulation.
* add a lambda and stream in TaskManagement.

This is preferable as users now know the actual task numbers corresponding to tasks displayed in search results.

yhtMinceraft1010X committed:
Add sort functionality

Tasks are displayed in the order in which they were added.

This is very inconvenient for keeping track of which tasks to do first.

Let's,
* add a priority level for tasks.
* add a way to sort tasks by priority and whether task was done.

This helps to keep track of which tasks to do first.

yhtMinceraft1010X committed:
Merge branch 'branch-BCD-Extension'

yhtMinceraft1010X committed:
Add more streams

yhtMinceraft1010X committed:
Add continuous integration

Let's add continuous integration to smoothen software engineering workflows.

yhtMinceraft1010X committed:
Expand edit functionality

Edit allows for the description of the task to be changed without deleting the task.

This is still inconvenient for changing date time and priority level.

Let's expand edit such that it can also change date time and/or priority level.

This allows for all details regarding a task to be edited without having to delete the task.

yhtMinceraft1010X committed:
Reformat arguments

The edit command for deadline and event tasks has format (new description) ; (new priority) ; (new datetime).
The commands to add above tasks are of format (description) ; (datetime) ; (priority).

This can confuse users.

Let's standardise the format to (description) ; (priority) ; (datetime).

This is easier to remember and consistent.

yhtMinceraft1010X committed:
Add new sorting criteria

Let's add a new criteria to sort tasks based on the lexicographical order of their descriptions.

yhtMinceraft1010X committed:
Update GUI and User Guide

Let's make some minor tweaks to the GUI and User Guide.

yhtMinceraft1010X committed:
Set theme jekyll-theme-slate

yhtMinceraft1010X committed:
Refactor task codes

Task codes like "T", "D" and "E" can be found in multiple classes.

This is cumbersome to change should the format of task code be altered.

Let's move the task code constants to a separate class.

This allows for one-stop changes to the task code formatting.

yhtMinceraft1010X committed:
Merge branch 'master' of https://github.com/yhtMinceraft1010X/ip-1

yhtMinceraft1010X committed:
Update build.gradle for release

Let's update the build.gradle file for correct generation of jar file.

yhtMinceraft1010X committed:
Update Javadoc

Let's add a blank line before tags for all javadoc comments.

yhtMinceraft1010X committed:
Update Checkstyle

The javadoc comments are manually edited to include blank line between description and tags.

This is tedious and some are missed out.

Let's,
* update checkstyle to 8.36.2.
* add new checkstyle module to check for empty line between description and tags.

This helps to automate checking of javadoc comments to more closely fit coding standards.

yhtMinceraft1010X committed:
Downsize GUI

GUI is large to maximise number of conversation lines without scrolling down.

Some laptops however have a smaller screen size and the GUI is unable to scale down.

Let's downsize the GUI.

This helps improve viewability of GUI.

yhtMinceraft1010X committed:
Update FXML

Let's change the FXML settings from JavaFX 15 to JavaFX 11.

yhtMinceraft1010X committed:
Merge branch 'master' of https://github.com/nus-cs2103-AY2021S2/ip into update-from-super-upstream

yhtMinceraft1010X committed:
Merge pull request #4 from yhtMinceraft1010X/update-from-super-upstream

Update from super upstream
